numpy/core/_internal.pyi

```diff
@@ -1,18 +1,37 @@
-from typing import Any
+from typing import Any, TypeVar, Type, overload, Optional, Generic
+import ctypes as ct
 
-# TODO: add better annotations when ctypes is stubbed out
+from numpy import ndarray
 
-class _ctypes:
+_CastT = TypeVar("_CastT", bound=ct._CanCastTo)  # Copied from `ctypes.cast`
+_CT = TypeVar("_CT", bound=ct._CData)
+_PT = TypeVar("_PT", bound=Optional[int])
+
+# TODO: Let the likes of `shape_as` and `strides_as` return `None`
+# for 0D arrays once we've got shape-support
+
+class _ctypes(Generic[_PT]):
+    @overload
+    def __new__(cls, array: ndarray[Any, Any], ptr: None = ...) -> _ctypes[None]: ...
+    @overload
+    def __new__(cls, array: ndarray[Any, Any], ptr: _PT) -> _ctypes[_PT]: ...
+
+    # NOTE: In practice `shape` and `strides` return one of the concrete
+    # platform dependant c_int64-based array-types (`c_int`, `c_long` or
+    # `c_longlong`)
```

Comment:
```suggestion
    # NOTE: In practice `shape` and `strides` return one of the concrete
    # platform dependant array-types (`c_int`, `c_long` or
    # `c_longlong`) corresponding to C's `int_ptr_t`
```
In particular, this can be 32-bit on 32-bit platforms.

Reply:
Well, that's most certainly annoying.
Comment updated as of https://github.com/numpy/numpy/pull/19029/commits/f2473b6605711f01b3f19e6eb6a1cdd7bb777c67.

Reply:
Just to note that `c_int` is 32 bits on all platforms that we currently support.